File: ortools/math_opt/python/indicator_constraints_test.py
#!/usr/bin/env python3
# Copyright 2010-2025 Google LLC
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math
from typing import Dict

from absl.testing import absltest
from ortools.math_opt.python import indicator_constraints
from ortools.math_opt.python import model
from ortools.math_opt.python import variables


def _terms_dict(
    con: indicator_constraints.IndicatorConstraint,
) -> Dict[variables.Variable, float]:
    return {term.variable: term.coefficient for term in con.terms()}


class IndicatorConstraintsTest(absltest.TestCase):

    def test_getters_empty(self) -> None:
        mod = model.Model()
        con = mod.add_indicator_constraint()
        self.assertIsNone(con.indicator_variable)
        self.assertEmpty(list(con.terms()))
        self.assertEqual(con.lower_bound, -math.inf)
        self.assertEqual(con.upper_bound, math.inf)
        self.assertFalse(con.activate_on_zero)
        self.assertEqual(con.name, "")
        bounded_expr = con.get_implied_constraint()
        self.assertEqual(bounded_expr.lower_bound, -math.inf)
        self.assertEqual(bounded_expr.upper_bound, math.inf)
        expr = variables.as_flat_linear_expression(bounded_expr.expression)
        self.assertEqual(expr.offset, 0.0)
        self.assertEmpty(expr.terms)

    def test_getters_nonempty(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        y = mod.add_variable()
        z = mod.add_variable()
        con = mod.add_indicator_constraint(
            indicator=x,
            activate_on_zero=True,
            implied_constraint=2 * y + z == 3.0,
            name="c123",
        )
        self.assertEqual(con.indicator_variable, x)
        self.assertDictEqual(_terms_dict(con), {y: 2.0, z: 1.0})
        self.assertEqual(con.lower_bound, 3.0)
        self.assertEqual(con.upper_bound, 3.0)
        self.assertTrue(con.activate_on_zero)
        self.assertEqual(con.name, "c123")
        self.assertEqual(con.get_coefficient(y), 2.0)
        self.assertEqual(con.get_coefficient(x), 0.0)

        bounded_expr = con.get_implied_constraint()
        self.assertEqual(bounded_expr.lower_bound, 3.0)
        self.assertEqual(bounded_expr.upper_bound, 3.0)
        expr = variables.as_flat_linear_expression(bounded_expr.expression)
        self.assertEqual(expr.offset, 0.0)
        self.assertEqual(expr.terms, {y: 2.0, z: 1.0})

    def test_create_by_attrs(self) -> None:
        mod = model.Model()
        x = mod.add_binary_variable()
        y = mod.add_variable()
        z = mod.add_variable()
        con = mod.add_indicator_constraint(
            indicator=x,
            activate_on_zero=True,
            implied_lb=4.0,
            implied_ub=5.0,
            implied_expr=10 * y + 9 * z + 3.0,
            name="c123",
        )
        self.assertEqual(con.indicator_variable, x)
        self.assertDictEqual(_terms_dict(con), {y: 10.0, z: 9.0})
        self.assertEqual(con.lower_bound, 1.0)
        self.assertEqual(con.upper_bound, 2.0)
        self.assertTrue(con.activate_on_zero)
        self.assertEqual(con.name, "c123")

    def test_get_coefficient_wrong_model(self) -> None:
        mod = model.Model()
        x = mod.add_variable()
        mod2 = model.Model()
        con = mod2.add_indicator_constraint()
        with self.assertRaises(ValueError):
            con.get_coefficient(x)

    def test_eq(self) -> None:
        mod_a = model.Model()
        con_a1 = mod_a.add_indicator_constraint()
        con_a2 = mod_a.add_indicator_constraint()
        con_a1_alt = mod_a.get_indicator_constraint(0)

        mod_b = model.Model()
        con_b1 = mod_b.add_indicator_constraint()

        self.assertEqual(con_a1, con_a1)
        self.assertEqual(con_a1, con_a1_alt)
        self.assertNotEqual(con_a1, con_a2)
        self.assertNotEqual(con_a1, con_b1)
        self.assertNotEqual(con_a1, "cat")

    def test_hash_no_crash(self) -> None:
        mod_a = model.Model()
        con = mod_a.add_indicator_constraint()
        self.assertIsInstance(hash(con), int)


if __name__ == "__main__":
    absltest.main()

File: ortools/javatests/com/google/ortools/mathopt/IndicatorConstraintsTest.java
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.ortools.mathopt;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.extensions.proto.ProtoTruth.assertThat;
import static com.google.ortools.mathopt.Expressions.linExpr;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.google.common.collect.ImmutableMap;
import com.google.ortools.mathopt.IndicatorConstraintUpdatesProto;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public final class IndicatorConstraintsTest {
  private ModelId modelId;
  private Variables variables;
  private IndicatorConstraints indicatorConstraints;

  @BeforeEach
  public void setUp() {
    modelId = new ModelId("model");
    variables = new Variables(modelId);
    indicatorConstraints = new IndicatorConstraints(modelId);
  }

  @Test
  public void addIndicatorConstraint_addsConstraint() {
    Variable z = variables.addVariable("z");
    Variable x = variables.addVariable("x");

    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 1.0, 2.0, linExpr(x).add(3.0), "c");

    assertThat(c.getId()).isEqualTo(0);
    assertThat(c.getName()).isEqualTo("c");
    assertThat(c.getIndicatorVariable()).isEqualTo(z);
    assertThat(c.getActivateOnZero()).isTrue();
    // Offset 3.0 subtracted from bounds
    assertThat(c.getLowerBound()).isEqualTo(-2.0);
    assertThat(c.getUpperBound()).isEqualTo(-1.0);
    assertThat(c.getUnmodifiableImpliedTerms()).containsExactly(x, 1.0);

    assertThat(indicatorConstraints.numIndicatorConstraints()).isEqualTo(1);
    assertThat(indicatorConstraints.hasIndicatorConstraintWithId(0)).isTrue();
    assertThat(indicatorConstraints.getIndicatorConstraint(0)).isEqualTo(c);
  }

  @Test
  public void deleteIndicatorConstraint_deletes() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");

    assertThat(indicatorConstraints.deleteIndicatorConstraint(c)).isTrue();

    assertThat(c.isDeleted()).isTrue();
    assertThat(indicatorConstraints.numIndicatorConstraints()).isEqualTo(0);
    assertThat(indicatorConstraints.hasIndicatorConstraintWithId(0)).isFalse();
  }

  @Test
  public void diff_tracksCreation() {
    Variable z = variables.addVariable("z");
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();

    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");

    IndicatorConstraintUpdatesProto indicatorConstraintUpdatesProto = diff.update();

    assertThat(indicatorConstraintUpdatesProto.getNewConstraintsMap())
        .containsExactly(0L, c.toProto());
  }

  @Test
  public void diff_tracksDeletion() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();

    indicatorConstraints.deleteIndicatorConstraint(c);

    IndicatorConstraintUpdatesProto indicatorConstraintUpdatesProto = diff.update();

    assertThat(indicatorConstraintUpdatesProto.getDeletedConstraintIdsList()).containsExactly(0L);
  }

  @Test
  public void diff_advance_clearsUpdates() {
    Variable z = variables.addVariable("z");
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();
    var unused = indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");

    diff.advance();

    IndicatorConstraintUpdatesProto indicatorConstraintUpdatesProto = diff.update();

    assertThat(indicatorConstraintUpdatesProto.getNewConstraintsMap()).isEmpty();
  }

  @Test
  public void diff_advance_clearsDeleted() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c1 =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c1");
    IndicatorConstraint c2 =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c2");
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();

    indicatorConstraints.deleteIndicatorConstraint(c1);
    diff.advance();
    indicatorConstraints.deleteIndicatorConstraint(c2);

    IndicatorConstraintUpdatesProto indicatorConstraintUpdatesProto = diff.update();
    assertThat(indicatorConstraintUpdatesProto.getDeletedConstraintIdsList())
        .containsExactly(c2.getId());
  }

  @Test
  public void diff_deleteNewConstraint_notTrackedAsDeleted() {
    Variable z = variables.addVariable("z");
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();

    // Add a constraint AFTER creating the diff
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");
    // Delete it
    indicatorConstraints.deleteIndicatorConstraint(c);

    IndicatorConstraintUpdatesProto indicatorConstraintUpdatesProto = diff.update();

    // Should not be in deletedConstraintIds because it was created after checkpoint
    assertThat(indicatorConstraintUpdatesProto.getDeletedConstraintIdsList()).isEmpty();
    // Should not be in newConstraints because it is deleted
    assertThat(indicatorConstraintUpdatesProto.getNewConstraintsMap()).isEmpty();
  }

  @Test
  public void variableDeletion_updatesIndicatorConstraint() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 0.0, 0.0, linExpr(), "c");

    indicatorConstraints.onVariableDeleted(z);

    assertThat(c.getIndicatorVariable()).isNull();
  }

  @Test
  public void deleteIndicatorConstraint_stopsTrackingIndicatorVariable() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 0.0, 0.0, linExpr(), "c");

    indicatorConstraints.deleteIndicatorConstraint(c);
    // Simulate model behavior:
    indicatorConstraints.onVariableDeleted(z);

    // If c was correctly removed from internal maps of indicatorConstraints, onVariableDeleted(z)
    // should not have found c, and thus should not have called c.onVariableDeleted(z).
    assertThat(c.getIndicatorVariable()).isEqualTo(z);
  }

  @Test
  public void deleteIndicatorConstraint_stopsTrackingImpliedTermVariable() {
    Variable z = variables.addVariable("z");
    Variable x = variables.addVariable("x");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 0.0, 0.0, linExpr(x), "c");

    indicatorConstraints.deleteIndicatorConstraint(c);
    // Simulate model behavior:
    indicatorConstraints.onVariableDeleted(x);

    // If c was correctly removed from internal maps of indicatorConstraints, onVariableDeleted(x)
    // should not have found c, and thus should not have called c.onVariableDeleted(x).
    // c.onVariableDeleted(x) would remove x from terms.
    assertThat(c.getUnmodifiableImpliedTerms()).containsKey(x);
  }

  @Test
  public void deleteIndicatorConstraintWithDeletedIndicator_stopsTrackingImpliedTermVariable() {
    Variable z = variables.addVariable("z");
    Variable x = variables.addVariable("x");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 0.0, 0.0, linExpr(x), "c");
    // Delete the indicator variable.
    variables.deleteVariable(z);
    indicatorConstraints.onVariableDeleted(z);

    // Delete the indicator constraint.
    indicatorConstraints.deleteIndicatorConstraint(c);
    indicatorConstraints.onVariableDeleted(x);

    // If c was correctly removed from internal maps of indicatorConstraints, onVariableDeleted(x)
    // should not have found c, and thus should not have called c.onVariableDeleted(x).
    // c.onVariableDeleted(x) would remove x from terms.
    assertThat(c.getUnmodifiableImpliedTerms()).containsKey(x);
  }

  @Test
  public void removeDiff_removesDiff() {
    IndicatorConstraints.Diff diff = indicatorConstraints.addDiff();
    indicatorConstraints.removeDiff(diff);
    // If we can't access diffs list, we verify behavior (e.g., adding constraint doesn't explode,
    // but diff is already invalid?)
    // Actually, Model.java handles removeTracker logic.
    // We can verify that calling removeDiff with a removed diff throws.
    assertThrows(IllegalArgumentException.class, () -> indicatorConstraints.removeDiff(diff));
  }

  @Test
  public void getNextId_returnsNextId() {
    assertThat(indicatorConstraints.getNextId()).isEqualTo(0);
    Variable z = variables.addVariable("z");
    var unused = indicatorConstraints.addIndicatorConstraint(z, false, 0.0, 0.0, linExpr(), "c");
    assertThat(indicatorConstraints.getNextId()).isEqualTo(1);
  }

  @Test
  public void ensureNextIdAtLeast_updatesId() {
    indicatorConstraints.ensureNextIdAtLeast(10);
    assertThat(indicatorConstraints.getNextId()).isEqualTo(10);
  }

  @Test
  public void deleteIndicatorConstraint_removesIndicatorVariableFromLookupMap() {
    Variable z = variables.addVariable("z");
    IndicatorConstraint c =
        indicatorConstraints.addIndicatorConstraint(z, true, 0.0, 0.0, linExpr(), "c");
    IndicatorConstraint constraintForLookup =
        new IndicatorConstraint(10, "c", modelId, z, true, 3.0, 4.0, ImmutableMap.of());

    indicatorConstraints.deleteIndicatorConstraint(c);
    // Simulate model behavior:
    indicatorConstraints.onVariableDeleted(z);

    NullPointerException e = assertThrows(NullPointerException.class,
        () -> indicatorConstraints.deleteIndicatorConstraint(constraintForLookup));
    assertThat(e).hasMessageThat().contains("Variable z not present in variableToConstraint");
  }
}
